target/linux/x86/base-files/lib/preinit/43_grow_rootfspart

```diff
@@ -0,0 +1,36 @@
+get_squashfs_size() {
+	local low high
+	low=$(hexdump -s 40 -n 4 -e '1/4 "%08x" "\n"' "$1")
+	high=$(hexdump -s 44 -n 4 -e '1/4 "%08x" "\n"' "$1")
+	echo $((0x$high$low))
+}
+
+run_grow_root_part() {
+	[ "$FAILSAFE" = "true" ] && return
+
+	local diskdev squashfs_magic
+	local SQUASHFS_ALIGN=$((0x10000))
+
+	. /lib/upgrade/common.sh
+
+	if export_bootdevice && export_partdevice diskdev 0; then
+		diskdev=/dev/$diskdev
+		case "$diskdev" in
+			*[0-9]) part2dev=${diskdev}p2;;
+			*)      part2dev=${diskdev}2;;
+		esac
+		[ -e "$part2dev" ] || return
+		squashfs_magic=$(dd bs=4 count=1 if="$part2dev" 2>/dev/null)
+		[ "$squashfs_magic" = "hsqs" ] || return
```

Comment:
this fails on ext4 implementations... suggest a subroutine/function once you ( and/or TO + CASE ) identify the FSTYPE, cascaded logix seems problematic here...

Reply:
This is intentional not executed for ext4 rootfs (although it will work as expected for squashfs+ext4). Growing part2 is only required at preinit if it will be used by an overlay filesystem that will be formatted a little bit after. I just know the squashfs case. Would it be anything else similar to squashfs+overlay case for x86 targets that I should consider?

ext4 should be expandable, even while in use. It should only require resize2fs installed. There is no advantage of doing it during pre-init as ext4 is already mounted.

The script grow_rootfspart will gladly expand part2 if execute manually, even refreshing partitions without a reboot:

root@OpenWrt:/# /usr/sbin/grow_rootfspart /dev/vda
grow_rootfspart: Growing /dev/vda partition 2 from 212992 to 433152 sectors
grow_rootfspart: Saving /dev/vda new MBR partition table...

I could save the user from installing a partitioner. However, the ext4 shipped with a ext4 flavor does not seem to like being expanded. It might be caused by a fine-tune optimization trying to reduce the ext4 disk usage:

```
root@OpenWrt:/# opkg install resize2fs
root@OpenWrt:/# resize2fs /dev/vda2 
resize2fs 1.45.6 (20-Mar-2020)
Filesystem at /dev/vda2 is mounted on /; on-line resizing required
old_desc_blocks = 1, new_desc_blocks = 1
[   60.123874] EXT4-fs (vda2): resizing filesystem from 32768 to 54144 blocks
[   60.126318] EXT4-fs warning (device vda2): reserve_backup_gdb:1035: reserved block 8 not at offset 7
[   60.128802] EXT4-fs warning (device vda2): ext4_resize_fs:2122: error (-22) occurred during file system resize
[   60.130514] EXT4-fs (vda2): resized filesystem to 32768
Performing an on-line resize of /dev/vda2 to 54144 (4k) blocks.
[   60.133087] EXT4-fs warning (device vda2): reserve_backup_gdb:1035: reserved block 8 not at offset 7
resize2fs: Invalid argument While trying to add group #1
```

As long as this persists, it is a dead end. If it could be expanded, I would simply run the `grow_rootfspart /dev/xxx2` and `resize2fs /dev/xxx2` after the system was up.

Reply:
ok... so in it's current form... you expect it to fail/end silently on ext4 implementations... ( the mentioned code is not silent produces errors )...

in this case... maybe clarify this in the pr description and/or consider some way to toggle this functionality, either as a package selection / Makefile@squasfs/f2fs/pre-init option etc.

Reply:
> ok... so in it's current form... you expect it to fail/end silently on ext4 implementations...
Yes

> in this case... maybe clarify this in the pr description and/or consider some way to toggle this functionality, either as a package selection / Makefile@squasfs/f2fs or pre-init option etc.

This feature is specific for x86 targets where storage normally is not a big issue. I could convert it to a package but it would be mostly useless for other targets. No other target appears to be using GPT (even less likely MBR) for booting.The preinit part is intentionally innocuous for ext4 rootfs but grow_rootfspart could still be usable.

It would probably use a image config value (dependent on x86) instead of a standalone package. I just don't know if it is worth it adding a new menu entry and extra image building logic. Anyway, I can do it without a problem. However, before that, I just need to know if there is interest to merge this PR.

Reply:
most sdcard images are also capable / compatible with this codeset ( rpi, bpi, r2s etc. )

another possibility is a "sysupgrade -E" ( extend ) style option that would be silently ignored on typical non-extendable/flash devices...

it's part of a larger push on such devices ( multiboot pr, gpt, efi etc. even dual-fw to some extent ) to extensify disk operations and carries with it complexities which overlap with typical logic... and bring compound implications when it comes to image consistency and predictability across the project... ( dual-fw logic @ advanced reboot app is virtually hardcoded and needs to be library-fied )

i'm unable to speak for merging... but I would like to see the majority of the options and functions discussed condensed and selectable within the build system... or via clear user hooks and parameters...

adrians upgrade json parameters are a nice clear window into one place some of these capabilities might be 'negotitated'/hooked into.... ( note: on a currently running device, there is no way to query "am I factory-squashfs" or "am I factory-ext4" the asu software for example just selects the first image returned by the server... you have an (on-device) fix for this... so it should be in common.sh or common-fulldisk.sh etc. etc. and/or be available via os-release or json call device partitions etc. etc. if it's in os-release/system from factory... then you can do away with alot of validation / determinism )

in short, would I like to see it merged? > yes... with user selectability and better library-fication...

not consolidating alot of this stuff now impedes extensification and leads to drama down the road...

my 2c